File: kotlin-node/karakum/src/jsMain/resources/plugins/convertHttpTypeParameter.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

function isInstanceType(node) {
    return (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "InstanceType"
    )
}

export default function (node, context, render) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (
        !sourceFileName.endsWith("http.d.ts")
        && !sourceFileName.endsWith("https.d.ts")
        && !sourceFileName.endsWith("http2.d.ts")
    ) return null

    if (
        ts.isTypeParameterDeclaration(node)
        && node.parent
        && (
            (
                ts.isClassDeclaration(node.parent)
                && node.parent.name?.text === "Server"
            )
            || (
                ts.isInterfaceDeclaration(node.parent)
                && (
                    node.parent.name.text === "ServerOptions"
                    || node.parent.name.text === "SecureServerOptions"
                    || node.parent.name.text === "Http2Server"
                    || node.parent.name.text === "Http2SecureServer"
                    || node.parent.name.text === "ServerHttp2Session"
                    || node.parent.name.text === "ServerSessionOptions"
                    || node.parent.name.text === "SecureServerSessionOptions"
                )
            )
            || (
                ts.isTypeAliasDeclaration(node.parent)
                && node.parent.name.text === "ServerOptions"
            )
            || (
                ts.isFunctionDeclaration(node.parent)
                || node.parent.name?.text === "createServer"
                || node.parent.name?.text === "createSecureServer"
                || node.parent.name?.text === "performServerHandshake"
            )
        )
    ) {
        if (node.name.text === "Request") {
            return `Request : IncomingMessage`
        }

        if (node.name.text === "Http1Request") {
            return `Http1Request : IncomingMessage`
        }

        if (node.name.text === "Http2Request") {
            return `Http2Request : Http2ServerRequest`
        }

        if (node.name.text === "Response") {
            return `Response : ServerResponse<*>`
        }

        if (node.name.text === "Http1Response") {
            return `Http1Response : ServerResponse<*>`
        }

        if (node.name.text === "Http2Response") {
            return `Http2Response : Http2ServerResponse<*>`
        }
    }

    if (
        ts.isTypeParameterDeclaration(node)
        && node.parent
        && ts.isTypeAliasDeclaration(node.parent)
        && node.parent.name.text === "RequestListener"
    ) {
        if (node.name.text === "Request") {
            return `Request /* : IncomingMessage */`
        }

        if (node.name.text === "Response") {
            return `Response /* : ServerResponse<*> */`
        }
    }

    if (
        isInstanceType(node)
    ) {
        return render(node.typeArguments[0])
    }

    if (
        ts.isIntersectionTypeNode(node)
        && node.types[0]
        && isInstanceType(node.types[0])
    ) {
        return render(node.types[0].typeArguments[0])
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "Request"
            || node.typeName.text === "Http1Request"
            || node.typeName.text === "Http2Request"

            || node.typeName.text === "Response"
            || node.typeName.text === "Http1Response"
            || node.typeName.text === "Http2Response"
        )
    ) {
        if (
            node.parent
            && ts.isTypeReferenceNode(node.parent)
            && (
                (
                    ts.isIdentifier(node.parent.typeName)
                    && (
                        node.parent.typeName.text === "InstanceType"

                        || node.parent.typeName.text === "Server"
                        || node.parent.typeName.text === "ServerOptions"
                        || node.parent.typeName.text === "RequestListener"

                        || node.parent.typeName.text === "Http2Server"
                        || node.parent.typeName.text === "ServerHttp2Session"
                    )
                )
                || (
                    ts.isQualifiedName(node.parent.typeName)
                    && (
                        node.parent.typeName.right.text === "ServerOptions"
                        || node.parent.typeName.right.text === "RequestListener"
                    )
                )
            )
        ) {
            return null
        }

        if (
            node.parent
            && ts.isExpressionWithTypeArguments(node.parent)
            && ts.isIdentifier(node.parent.expression)
            && (
                node.parent.expression.text === "ServerSessionOptions"
                || node.parent.expression.text === "SecureServerSessionOptions"
            )
        ) {
            return null
        }

        const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
        const typeChecker = typeScriptService?.program.getTypeChecker()

        const symbol = typeChecker?.getSymbolAtLocation(node.typeName)

        const typeParameterDeclarations = (symbol?.declarations ?? [])
        const typeParameterDeclaration = typeParameterDeclarations
            .find(declaration => ts.isTypeParameterDeclaration(declaration))

        if (
            typeParameterDeclaration
            && typeParameterDeclaration.parent
            && (
                (
                    ts.isClassDeclaration(typeParameterDeclaration.parent)
                    && typeParameterDeclaration.parent.name?.text === "Server"
                )
                || (
                    ts.isInterfaceDeclaration(typeParameterDeclaration.parent)
                    && (
                        typeParameterDeclaration.parent.name.text === "ServerOptions"
                        || typeParameterDeclaration.parent.name.text === "SecureServerOptions"
                        || typeParameterDeclaration.parent.name.text === "Http2Server"
                        || typeParameterDeclaration.parent.name.text === "Http2SecureServer"
                        || typeParameterDeclaration.parent.name.text === "ServerHttp2Session"
                        || typeParameterDeclaration.parent.name.text === "ServerSessionOptions"
                        || typeParameterDeclaration.parent.name.text === "SecureServerSessionOptions"
                    )
                )
                || (
                    ts.isTypeAliasDeclaration(typeParameterDeclaration.parent)
                    && typeParameterDeclaration.parent.name.text === "ServerOptions"
                )
                || (
                    ts.isFunctionDeclaration(typeParameterDeclaration.parent)
                    && typeParameterDeclaration.parent.name?.text === "createServer"
                )
                || (
                    ts.isTypeAliasDeclaration(typeParameterDeclaration.parent)
                    && typeParameterDeclaration.parent.name.text === "RequestListener"
                )
            )
        ) {
            return `JsClass<${render(node)}>`
        }
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertHttpTypeParameter.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.Node
import typescript.TypeReferenceNode
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isExpressionWithTypeArguments
import typescript.isFunctionDeclaration
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isIntersectionTypeNode
import typescript.isQualifiedName
import typescript.isTypeAliasDeclaration
import typescript.isTypeParameterDeclaration
import typescript.isTypeReferenceNode
import kotlin.contracts.ExperimentalContracts
import kotlin.contracts.contract

@OptIn(ExperimentalContracts::class)
@Suppress("CANNOT_CHECK_FOR_EXTERNAL_INTERFACE", "ERROR_IN_CONTRACT_DESCRIPTION")
private fun isInstanceType(node: Node): Boolean {
    contract {
        returns(true) implies (node is TypeReferenceNode)
    }

    return nullable {
        ensure(isTypeReferenceNode(node))

        val name = node.typeName
        ensure(isIdentifier(name))
        ensure(name.text == "InstanceType")
    } != null
}

private fun isTypeArgumentOfTypeWithTypeofBound(node: Node) = nullable {
    nullable {
        val typeReference = ensureNotNull(node.getParentOrNull())
        ensure(isTypeReferenceNode(typeReference))

        nullable {
            val typeName = typeReference.typeName
            ensure(isIdentifier(typeName))
            ensure(
                typeName.text == "InstanceType"

                        || typeName.text == "Server"
                        || typeName.text == "ServerOptions"
                        || typeName.text == "SecureServerOptions"
                        || typeName.text == "RequestListener"

                        || typeName.text == "Http2Server"
                        || typeName.text == "Http2SecureServer"
                        || typeName.text == "ServerHttp2Session"
            )
        } ?: nullable {
            val typeName = typeReference.typeName
            ensure(isQualifiedName(typeName))
            ensure(
                typeName.right.text == "ServerOptions"
                        || typeName.right.text == "RequestListener"
            )
        }
    } ?: nullable {
        val expressionWithTypeArguments = ensureNotNull(node.getParentOrNull())
        ensure(isExpressionWithTypeArguments(expressionWithTypeArguments))

        val expression = expressionWithTypeArguments.expression
        ensure(isIdentifier(expression))
        ensure(
            expression.text == "ServerSessionOptions"
                    || expression.text == "SecureServerSessionOptions"
        )
    }
} != null

val convertHttpTypeParameter = createPlugin { node, context, render ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName

        ensure(
            sourceFileName.endsWith("http.d.ts")
                    || sourceFileName.endsWith("https.d.ts")
                    || sourceFileName.endsWith("http2.d.ts")
        )

        nullable {
            ensure(isTypeParameterDeclaration(node))

            ensureNotNull(
                nullable {
                    val classNode = ensureNotNull(node.getParentOrNull())
                    ensure(isClassDeclaration(classNode))
                    ensure(classNode.name?.text == "Server")
                } ?: nullable {
                    val interfaceNode = ensureNotNull(node.getParentOrNull())
                    ensure(isInterfaceDeclaration(interfaceNode))
                    ensure(
                        interfaceNode.name.text == "Server"
                                || interfaceNode.name.text == "ServerOptions"
                                || interfaceNode.name.text == "SecureServerOptions"
                                || interfaceNode.name.text == "Http2Server"
                                || interfaceNode.name.text == "Http2SecureServer"
                                || interfaceNode.name.text == "ServerHttp2Session"
                                || interfaceNode.name.text == "ServerSessionOptions"
                                || interfaceNode.name.text == "SecureServerSessionOptions"
                    )
                } ?: nullable {
                    val typeAlias = ensureNotNull(node.getParentOrNull())
                    ensure(isTypeAliasDeclaration(typeAlias))
                    ensure(typeAlias.name.text == "ServerOptions")
                } ?: nullable {
                    val function = ensureNotNull(node.getParentOrNull())
                    ensure(isFunctionDeclaration(function))
                    ensure(
                        function.name?.text == "createServer"
                                || function.name?.text == "createSecureServer"
                                || function.name?.text == "performServerHandshake"
                    )
                }
            )

            when {
                node.name.text == "Request" -> "Request : IncomingMessage"
                node.name.text == "Http1Request" -> "Http1Request : IncomingMessage"
                node.name.text == "Http2Request" -> "Http2Request : Http2ServerRequest"
                node.name.text == "Response" -> "Response : ServerResponse<*>"
                node.name.text == "Http1Response" -> "Http1Response : ServerResponse<*>"
                node.name.text == "Http2Response" -> "Http2Response : Http2ServerResponse<*>"
                else -> null
            }
        } ?: nullable {
            ensure(isTypeParameterDeclaration(node))

            val typeAlias = ensureNotNull(node.getParentOrNull())
            ensure(isTypeAliasDeclaration(typeAlias))
            ensure(typeAlias.name.text == "RequestListener")

            when {
                node.name.text == "Request" -> "Request /* : IncomingMessage */"
                node.name.text == "Response" -> "Response /* : ServerResponse<*> */"
                else -> null
            }
        } ?: nullable {
            ensure(isInstanceType(node))

            val firstTypeArgument = ensureNotNull(node.typeArguments?.asArray()?.getOrNull(0))

            render(firstTypeArgument)
        } ?: nullable {
            ensure(isIntersectionTypeNode(node))

            val firstType = ensureNotNull(node.types.asArray().getOrNull(0))
            ensure(isInstanceType(firstType))

            val firstTypeArgument = ensureNotNull(firstType.typeArguments?.asArray()?.getOrNull(0))

            render(firstTypeArgument)
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val name = node.typeName
            ensure(isIdentifier(name))
            ensure(
                name.text == "Request"
                        || name.text == "Http1Request"
                        || name.text == "Http2Request"

                        || name.text == "Response"
                        || name.text == "Http1Response"
                        || name.text == "Http2Response"
            )

            ensure(!isTypeArgumentOfTypeWithTypeofBound(node))

            nullable {
                val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))
                val typeChecker = typeScriptService.program.getTypeChecker()

                val symbol = ensureNotNull(typeChecker.getSymbolAtLocation(name))

                val typeParameterDeclarations = ensureNotNull(symbol.declarations)
                val typeParameterDeclaration = ensureNotNull(typeParameterDeclarations.find { isTypeParameterDeclaration(it) })

                ensureNotNull(
                    nullable {
                        val classNode = ensureNotNull(typeParameterDeclaration.getParentOrNull())
                        ensure(isClassDeclaration(classNode))
                        ensure(classNode.name?.text == "Server")
                    } ?: nullable {
                        val interfaceNode = ensureNotNull(typeParameterDeclaration.getParentOrNull())
                        ensure(isInterfaceDeclaration(interfaceNode))
                        ensure(
                            interfaceNode.name.text == "ServerOptions"
                                    || interfaceNode.name.text == "SecureServerOptions"
                                    || interfaceNode.name.text == "Http2Server"
                                    || interfaceNode.name.text == "Http2SecureServer"
                                    || interfaceNode.name.text == "ServerHttp2Session"
                                    || interfaceNode.name.text == "ServerSessionOptions"
                                    || interfaceNode.name.text == "SecureServerSessionOptions"
                        )
                    } ?: nullable {
                        val typeAlias = ensureNotNull(typeParameterDeclaration.getParentOrNull())
                        ensure(isTypeAliasDeclaration(typeAlias))
                        ensure(
                            typeAlias.name.text == "ServerOptions"
                                || typeAlias.name.text == "RequestListener"
                        )
                    } ?: nullable {
                        val function = ensureNotNull(typeParameterDeclaration.getParentOrNull())
                        ensure(isFunctionDeclaration(function))
                        ensure(
                            function.name?.text == "createServer"
                                    || function.name?.text == "createSecureServer"
                        )
                    }
                )

                "JsClass<${render(node)}>"
            }
        }
    }
}
